### PR TITLE
Feature/smoothed tv prior

### PR DESCRIPTION
**Pull request by Kushagra481**
Thank you for contributing to DeepInverse!

Please read our [contributing guidelines](https://deepinv.org/contributing.html) for instructions on how to contribute, including writing tests, documentation and code style.

Once the GitHub tests have been approved by a maintainer (only required for first-time contributors), and the `Build Docs` GitHub action has run successfully, you can download the documentation as a zip file from the [Actions page](https://github.com/deepinv/deepinv/actions/workflows/docs_cpu.yml). Look for the workflow run corresponding to your pull request.

### Checks to be done before submitting your PR

- [ ] `python3 -m pytest deepinv/tests` runs successfully.
- [ ] `black .` and `ruff check .` run successfully.
- [ ] `make html` runs successfully (in the `docs/` directory).
- [ ] Updated docstrings related to the changes (as applicable).
- [ ] Added an entry to the [changelog.rst](https://github.com/deepinv/deepinv/blob/main/docs/source/changelog.rst).

### LLM policy

LLM usage is ok, but not PRs generated 100% by AI. See our [LLM policy](https://deepinv.org/contributing.html#llm-policy) Tick below as appropriate:

- [ ] I did not use LLM tools to write the code
- [ ] LLM tools helped me to write part of the code
- [ ] An LLM tool wrote all of the code.
- [ ] An agent submitted the PR and wrote the description.

### CODE REVIEW

**Comment by Kushagra481** [comment]
#1260 issue started PR created a unit test and implementation was confused about the eps so would love some help @Tmodrzyk

**Comment by Kushagra481** [comment]
welp currently the fail is in the pipeline so please check if any changes from your side I'll fix @Tmodrzyk

**Comment by Tmodrzyk** [comment]
Hi @Kushagra481, we should probably inherit from the `TVPrio` class to avoid code duplication here. 
The tests should be included within the general ones, and not in a separate stand-alone function.

**Comment by Kushagra481** [comment]
@Tmodrzyk convergence is failing when I add SmoothedTVPrior class so should I try changing the code thres_conv or max_iter cuz its not converging like I can change thres_conv to a variable value that if smooth prior activates then use that value ? is this fine or am I missing something here

**Comment by Tmodrzyk** [comment]
Hi @Kushagra481, the class should inherit from `TVPrior`, it is still not the case: you redefined `.nabla` etc.
I think the test is failing because you set the default smoothing `eps=1e-3` which is fairly large, we can try smaller values. :smile:

**Comment by Kushagra481** [comment]
@Tmodrzyk any changes

**Comment by Tmodrzyk** [comment]
I'll do a proper review later next week, I'mvery busy at the moment :)

**Comment by Kushagra481** [comment]
Hey @Tmodrzyk, If you're free, could you please take a look at my PR when you get a chance? Just a gentle reminder in case it slipped your mind. Thanks a lot!

File: deepinv/optim/prior.py [hunk 1]
```diff
@@ -542,6 +542,83 @@ def nabla_adjoint(self, x: torch.Tensor) -> torch.Tensor:
...
 
+class SmoothedTVPrior(TVPrior):
+    r"""
+    Smoothed total variation (TV) prior :math:`\reg{x} = \sum_i \sqrt{\|(Dx)_i\|_2^2 + \varepsilon^2}`.
+
+    This is a differentiable relaxation of the isotropic TV prior :class:`~deepinv.optim.TVPrior`, obtained
+    by replacing the (non-differentiable at 0) 2-norm in each pixel's finite-difference vector by a smoothed
+    version. As :math:`\varepsilon \to 0`, the smoothed prior converges to the standard TV prior. Unlike
+    :class:`~deepinv.optim.TVPrior`, which is implicit and only exposes a proximal operator, this prior is
+    explicit and differentiable everywhere, with closed-form gradient
```
Comment (Tmodrzyk) [inline deepinv/optim/prior.py]:
```suggestion
    :class:`~deepinv.optim.TVPrior`, which is non-differentiable and not guaranteed to be stable when using `.grad()`, this prior is differentiable everywhere.
    It does not rely on PyTorch's autodifferientation is is implemented explicitly.
```

File: deepinv/optim/prior.py [hunk 2]
```diff
@@ -542,6 +542,83 @@ def nabla_adjoint(self, x: torch.Tensor) -> torch.Tensor:
...
+        \nabla \reg{x} = D^\top \left( \frac{Dx}{\sqrt{\|Dx\|_2^2 + \varepsilon^2}} \right)
+
+    where :math:`D` is the finite differences linear operator, :math:`D^\top = -\operatorname{div}` its adjoint,
+    and the 2-norm is taken over the dimension of the differences at each pixel. This is useful for algorithms
+    that require an explicit, smooth gradient (e.g. gradient descent) rather than a proximal operator, and is
+    commonly used in emission tomography.
+
+    :param float eps: smoothing parameter :math:`\varepsilon > 0`. Larger values yield a smoother, more
+        strongly convexified approximation of TV; smaller values approach the nonsmooth TV prior more
+        closely. Default: 1e-3.
```
Comment (Tmodrzyk) [inline deepinv/optim/prior.py]:
```suggestion
        closely. Default: 1e-5.
```

File: deepinv/optim/prior.py [hunk 3]
```diff
@@ -542,6 +542,83 @@ def nabla_adjoint(self, x: torch.Tensor) -> torch.Tensor:
...
+
+        \nabla \reg{x} = D^\top \left( \frac{Dx}{\sqrt{\|Dx\|_2^2 + \varepsilon^2}} \right)
+
+    where :math:`D` is the finite differences linear operator, :math:`D^\top = -\operatorname{div}` its adjoint,
+    and the 2-norm is taken over the dimension of the differences at each pixel. This is useful for algorithms
+    that require an explicit, smooth gradient (e.g. gradient descent) rather than a proximal operator, and is
+    commonly used in emission tomography.
+
+    :param float eps: smoothing parameter :math:`\varepsilon > 0`. Larger values yield a smoother, more
+        strongly convexified approximation of TV; smaller values approach the nonsmooth TV prior more
```
Comment (Tmodrzyk) [inline deepinv/optim/prior.py]:
```suggestion
    :param float eps: smoothing parameter :math:`\varepsilon > 0`. Larger values yield a smoother approximation of TV; smaller values approach the nonsmooth TV prior more
```

File: deepinv/optim/prior.py [hunk 4]
```diff
@@ -542,6 +542,83 @@ def nabla_adjoint(self, x: torch.Tensor) -> torch.Tensor:
...
+    explicit and differentiable everywhere, with closed-form gradient
+
+    .. math::
+
+        \nabla \reg{x} = D^\top \left( \frac{Dx}{\sqrt{\|Dx\|_2^2 + \varepsilon^2}} \right)
+
+    where :math:`D` is the finite differences linear operator, :math:`D^\top = -\operatorname{div}` its adjoint,
+    and the 2-norm is taken over the dimension of the differences at each pixel. This is useful for algorithms
+    that require an explicit, smooth gradient (e.g. gradient descent) rather than a proximal operator, and is
+    commonly used in emission tomography.
```
Comment (Tmodrzyk) [inline deepinv/optim/prior.py]:
```suggestion
    and the 2-norm is taken over the dimension of the differences at each pixel. 
```

File: deepinv/tests/test_optim.py [hunk 5]
```diff
@@ -615,6 +619,16 @@ def test_priors_algo(pnp_algo, imsize, dummy_dataset, device):
...
         prior = get_prior(prior_name, device=device)
+        if prior_name == "SmoothedTVPrior":
+            with torch.enable_grad():
+                x_ag = test_sample.clone().double().requires_grad_(True)
+                g = prior.fn(x_ag).sum()
+                (autograd_grad,) = torch.autograd.grad(g, x_ag)
+            closed_form_grad = prior.grad(test_sample.double())
+            assert torch.allclose(autograd_grad, closed_form_grad, atol=1e-8)
+            with pytest.raises(ValueError):
+                dinv.optim.prior.SmoothedTVPrior(eps=0.0)
```
Comment (Tmodrzyk) [inline deepinv/tests/test_optim.py]: This is a nice test, but it will unfortunately clutter this test function. We can remove it for now and think later about broader tests comparing autodiff vs explicit gradients. 😄

**Review by Tmodrzyk — CHANGES_REQUESTED** [review]
~~Hi @kushagra481, we're almost there ! A few changes in the docstring and we can remove the autograd vs explicit part of the test.
I'll give this a final look after these changes. :)~~
Oops, I missed something: the proximal operator is inherited from `TVPrior` now, so the new class will not compute the actual prox of the `SmoothedTVPrior`.
A quick fix and the easiest for you would be to implement the prox and raise a `NotImplementedError`. Handling the new prox computation is not trivial and I'm not sure it's worth spending time on this since the whole point of this class is to use the gradient.

**Comment by Kushagra481** [comment]
@Tmodrzyk should i create function for it and then raise the error or some other way because I cant understand how to raise the error 
```
def prox(self, x, *args, gamma=1.0, **kwargs):
    raise NotImplementedError("Use .grad() instead — prox isn't implemented for this class.")``
```
like this

**Comment by Tmodrzyk** [comment]
Hi @Kushagra481, yes this is the correct way of doing it !

File: deepinv/tests/test_optim.py (hunk 5 above)
Comment (Kushagra481) [reply deepinv/tests/test_optim.py]: so should I not add a test at all from smoothed prior in the test_optim.py file?

**Comment by Kushagra481** [comment]
@Tmodrzyk I just merged into main , things failing any idea I tried to read logs but like TVL1 prior failing , also sorry for disturbing you so much you can reply when you get free ik you writing your phd thesis

**Comment by Tmodrzyk** [comment]
Hi @Kushagra481, I think the merge has gone wrong.
Looking at the current diffs in your PR (https://github.com/deepinv/deepinv/pull/1290/changes), it seems you've changed the implementation of `TVL1` during your merge. 
You should revert to the previous commit and re-attempt a merge.

**Comment by Kushagra481** [comment]
@Tmodrzyk sorry for not responding for the past few days. I didn't intentionally change the implementation of anything, I just merged `main` into my branch.
I'll revert to the previous commit and re-attempt the merge to make sure the diff only contains my intended changes. Should I rebase onto `main` instead, or just pull/merge `main` again?

**Comment by Tmodrzyk** [comment]
Hi @Kushagra481 no worries, there's no rush. The correct fix is to revert the merge, which apparently went wrong, meaning reverting to the previous commit, then pulling from main and resolving the merge conflict correctly.
It can be tricky if you never resolved merge conflicts but most IDEs have features to simplify this step. :smile: 

Note: you probably accepted incoming changes instead of a mix of your changes and the incoming ones

**Comment by Kushagra481** [comment]
I made a mistake by force-pushing the branch earlier, which resulted in the previous PR being closed and introduced a lot of unwanted changes. I honestly didn't expect it to cause that much of a mess 😭 Guess actions do have consequences. I've recreated the branch and will open a new PR with the clean changes. Sorry for the confusion!
@Tmodrzyk

**Comment by Tmodrzyk** [comment]
Hi @Kushagra481 no problem, that's very common when learning to use git ! Since the changes weren't that big the easiest option is indeed to start again in a new PR :smile:
